src/ompl/base/spaces/src/VanaStateSpace.cpp

```diff
@@ -0,0 +1,301 @@
+        {
+        }
+
+        unsigned int getDimension() const override
+        {
+            return 3;
+        }
+
+        void defaultCellSizes() override
+        {
+            cellSizes_.resize(3);
+            bounds_ = space_->as<VanaStateSpace>()->getBounds();
+            bounds_.resize(3);
+            cellSizes_[0] = (bounds_.high[0] - bounds_.low[0]) / magic::PROJECTION_DIMENSION_SPLITS;
+            cellSizes_[1] = (bounds_.high[1] - bounds_.low[1]) / magic::PROJECTION_DIMENSION_SPLITS;
+            cellSizes_[2] = (bounds_.high[2] - bounds_.low[2]) / magic::PROJECTION_DIMENSION_SPLITS;
+        }
+
+        void project(const State *state, Eigen::Ref<Eigen::VectorXd> projection) const override
+        {
+            projection = Eigen::Map<const Eigen::VectorXd>(
+                state->as<StateType>()->as<RealVectorStateSpace::StateType>(0)->values, 3);
+        }
+    };
+
+    registerDefaultProjection(std::make_shared<VanaDefaultProjection>(this));
+}
+
+DubinsStateSpace::StateType *VanaStateSpace::get2DPose(double x, double y, double yaw) const
+{
+    auto state = dubinsSpace_.allocState()->as<DubinsStateSpace::StateType>();
+    state->setXY(x, y);
+    state->setYaw(yaw);
+    return state;
+}
+
+bool VanaStateSpace::decoupled(const State *state1, const State *state2, double radius, PathType &result,
+                               DubinsStateSpace::StateType *endSZ) const
+{
+    result.verticalRadius_ = 1. / std::sqrt(1. / (rho_ * rho_) - 1. / (radius * radius));
+    if (!std::isfinite(result.verticalRadius_))
+    {
+        return false;
+    }
+
+    auto s1 = state1->as<StateType>();
+    auto s2 = state2->as<StateType>();
+
+    result.horizontalRadius_ = radius;
+    // note that we are exploiting properties of the memory layout of state types
+    result.pathXY_ = dubinsSpace_.dubins(state1, state2, radius);
+
+    result.startSZ_->setXY(0., (*s1)[2]);
+    result.startSZ_->setYaw(s1->pitch());
+    endSZ->setXY(radius * result.pathXY_.length(), (*s2)[2]);
+    endSZ->setYaw(s2->pitch());
+    result.pathSZ_ = dubinsSpace_.dubins(result.startSZ_, endSZ, result.verticalRadius_);
+
+    // in three cases the result is invalid:
+    // 1. path of type CCC (i.e., RLR or LRL)
+    // 2. pitch smaller than minPitch_
+    // 3. pitch greater than maxPitch_
+    if ((result.pathSZ_.type_[1] != DubinsStateSpace::DUBINS_STRAIGHT) ||
+        (result.pathSZ_.type_[0] == DubinsStateSpace::DUBINS_RIGHT &&
+         s1->pitch() - result.pathSZ_.length_[0] < minPitch_) ||
+        (result.pathSZ_.type_[0] == DubinsStateSpace::DUBINS_LEFT &&
+         s1->pitch() + result.pathSZ_.length_[0] > maxPitch_))
+    {
+        return false;
+    }
+
+    return true;
+}
+
+std::optional<VanaStateSpace::PathType> VanaStateSpace::getPath(const State *state1, const State *state2) const
+{
+    double radiusMultiplier = 2.;
+    auto scratch = dubinsSpace_.allocState()->as<DubinsStateSpace::StateType>();
```

Comment:
allocation in hotpath. Perhaps the PathType below can be moved to class variables to avoid the allocation in their ctors as well?

Reply:
that would violate thread safety